File: Graphs/maxArea.cpp
#include <bits/stdc++.h>
using namespace std;
int maxAreaOfIsland(vector<vector<int>> &grid)
{
    int row = grid.size();
    int col = grid[0].size();
    int maxArea = 0;
    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            if (grid[i][j] == 1)
            {

                int currArea = 0;
                dfs(grid, i, j, row, col, currArea);

                maxArea = max(maxArea, currArea);
            }
        }
    }

    return maxArea;
}

void dfs(vector<vector<int>> &grid, int i, int j, int row, int col, int &result)
{
    if (i < 0 || j < 0 || i >= row || j >= col || grid[i][j] == 0)
    {
        return;
    }

    grid[i][j] = 0;

    result++;

    dfs(grid, i + 1, j, row, col, result);
    dfs(grid, i - 1, j, row, col, result);
    dfs(grid, i, j + 1, row, col, result);
    dfs(grid, i, j - 1, row, col, result);
}

File: Graphs/maxArea.py
from typing import List


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        row, col = len(grid), len(grid[0])
        globalMax = 0

        def dfs(i, j):
            if i < 0 or j < 0 or i >= row or j >= col or grid[i][j] != 1:
                return

            grid[i][j] = 0
            self.localMax += 1

            dfs(i + 1, j)
            dfs(i - 1, j)
            dfs(i, j + 1)
            dfs(i, j - 1)

        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1:
                    self.localMax = 0
                    dfs(i, j)
                    globalMax = max(globalMax, self.localMax)

        return globalMax
